cpt-replica.rst

```diff
@@ -68,30 +68,52 @@ We suggest the following key parameters:
-    ``semanage port -m -t http_port_t -p tcp 8000``.  The ``cvmfs-server``
-    RPM for EL7 executes this command as a post-installation script.
+    Apache instead of a separate machine, to reduce the number of points
+    of failure. In that case caching can be disabled for the data (since
+    there's no need to store it again on the same disk), but caching is
+    helpful for the responses to geo api calls. Using a squid is also
+    helpful for participating in shared monitoring such as the `WLCG
+    Squid Monitor <http://wlcg-squid-monitor.cern.ch>`.
+    
+    Alternatively, separate Squid server machines may be configured in a
+    round-robin DNS and each forward to the Apache server, but note that
+    if any of them are down the entire service will be considered down
+    by CernVM-FS clients.  A front end hardware load balancer that
+    quickly takes a machine that is down out of service would help
+    reduce the impact.
+
+**High availability**
+    On the subject of availability, note that it is not advised to use
+    two separate complete Stratum 1 servers in a single round-robin
+    service because they will be updated at different rates.  That would
+    cause errors when a client sees an updated catalog from one Stratum
+    1 but tries to read corresponding data files from the other that does
+    not yet have the files.  Different Stratum 1s should either be
+    separately configured on the clients, or a pair can be configured as
+    a high availability active/standby pair using the cvmfs-contrib
+    `cvmfs-hastratum1 package <https://github.com/cvmfs-contrib/cvmfs-hastratum1>`.
+    An active/standby pair can also be managed by switching a DNS name
+    between two different servers.
 
 **DNS cache**
-    A Stratum 1 does a lot of DNS lookups, so we recommend installing a
-    DNS caching mechanism on the machine such as ``dnsmasq`` or
-    ``bind``. We do not recommend ``nscd`` since it does not honor the
-    DNS Time-To-Live protocol.
+    The geo api on a Stratum 1 does DNS lookups.  It caches lookups
+    for 5 minutes so the DNS server load does not tend to be severe, but
+    we still recommend installing a DNS caching mechanism on the machine
+    such as ``dnsmasq`` or ``bind``.  We do not recommend ``nscd`` since
+    it does not honor the DNS Time-To-Live protocol.  
 
 Squid Configuration
 -------------------
 
+If you participate in the Open Science Grid (OSG) or the European Grid
+Infrastructure (EGI), you are encouraged to use their distribution of
+squid called frontier-squid.  It is kept up to date with the latest
+squid bug fixes and has features for easier upgrading and monitoring.
+Step-by-step instructions for setting it up with a Stratum 1 is
+available in the `OSG documentation
+https://opensciencegrid.org/docs/other/install-cvmfs-stratum1/#configuring-frontier-squid`.
+
+Otherwise, a `squid` package is available in most Linux operating systems.
```

Comment:
How about adding "Other HTTP caching proxies could also be used."  ?
Some sites use e.g. nginx, ATS based on their preferences. Though squid is most commonly used I don't think it is specifically required for CVMFS. If there are any pros/cons regarding choice of proxy, perhaps that could also be described in some docs.

Reply:
In theory that's true, although I am unaware of anybody having tried it and so we can't recommend any configuration.  I think it's better to not mention it until somebody has proved how to do it.

Reply:
I think the sites using nginx/ATS have proved that it can be done, with just a standard HTTP forward caching proxy config.

Reply:
Do you know of sites that are doing it for CVMFS Stratum zeros?   If so, let's get their configuration in just as we have the configuration for squid.

Reply:
Ah no, I missed that this section is specifically about reverse proxies, not forward proxies.

Reply:
Ok, although in fact reverse proxies are the only case that other popular caching proxies support.  Squid is the only forward proxy in popular use that I am aware of.

Reply:
@DrDaveD I'm not sure what you mean. Forward caching proxying is implemented by several proxies, and we have sites using nginx and ATS forward caching proxies.

ATS: https://docs.trafficserver.apache.org/en/latest/admin-guide/configuration/transparent-forward-proxying.en.html
```
The Apache Traffic Server is a general purpose proxy, configurable as both a reverse and forward proxy. 
A forward proxy can be ... combined with a cache to lower your overall bandwidth usage.
```
nginx is very versatile in configuration. It can be used as a caching forward or reverse proxy (or redirector, or http server, etc.). Although the original design of nginx was perhaps more oriented towards reverse proxying, and it is perhaps less well known as a forward proxy, overall nginx is certainly far more widely used in the tech world than squid.
e.g.
https://ef.gy/using-nginx-as-a-proxy-server

Reply:
Those are really just toy configurations, not for shared production use.   The apache [security considerations](https://docs.trafficserver.apache.org/en/latest/admin-guide/configuration/transparent-forward-proxying.en.html#security-considerations) section says it is an open proxy which should not be used in practice.  The nginx configuration is only for running on localhost.  I agree nginx is very popular, but it is only popular as a reverse proxy.

Reply:
Okay, good to know ATS doesn't have the same ACL functionality as squid and one should use e.g. iptables to restrict incoming connections, but that doesn't mean it can't be used as a forward caching proxy, though if restricting the destination is important to one's security practices that would be a reason to use squid instead.

nginx forward proxying is not limited to localhost. It works fine as a forward proxy, there are a bunch of examples showing how to do it but it basically amounts to just setting `proxy_pass http://$http_host$uri$is_args$args;` .

Reply:
Right, that's the destination, but all the examples show running nginx on localhost, not as a general purpose server used by lots of clients.  The last comment on this [stackoverflow question](https://stackoverflow.com/questions/46060028/how-to-use-nginx-as-forward-proxy-for-any-requested-location) about it says it didn't work, and nobody responded.

Reply:
You mean the comment about not working for HTTPS? Yes that may be a limitation with using nginx as a forward proxy, though not relevant for use with CVMFS.

Reply:
No, I meant the very last comment referencing http://forward-proxy-server-ip:8888 but now I'm not sure that's relevant.  I don't see any examples of people talking about http_proxy= some other server than localhost.